### ISSUE TITLE
Better FlexibleItemAnimator

### ISSUE THREAD

**Issue by qihongbo** [davideas/FlexibleAdapter#570]
``` java
package android.support.v7.widget;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.View;

import eu.davidea.flexibleadapter.common.FlexibleItemAnimator;

public class FlexibleItemAnimatorEx extends FlexibleItemAnimator {

    @Override
    public boolean animateDisappearance(@NonNull ViewHolder viewHolder,
                                        @NonNull ItemHolderInfo preLayoutInfo, @Nullable ItemHolderInfo postLayoutInfo) {

        int oldLeft = preLayoutInfo.left;
        int oldTop = preLayoutInfo.top;
        View disappearingItemView = viewHolder.itemView;
        int newLeft = postLayoutInfo == null ? disappearingItemView.getLeft() : postLayoutInfo.left;
        int newTop = postLayoutInfo == null ? disappearingItemView.getTop() : postLayoutInfo.top;
        if (!viewHolder.isRemoved() && (oldLeft != newLeft || oldTop != newTop)) {
            disappearingItemView.layout(newLeft, newTop,
                    newLeft + disappearingItemView.getWidth(),
                    newTop + disappearingItemView.getHeight());

            return animateMove(viewHolder, oldLeft, oldTop, newLeft, newTop);
        } else {
            adjustRemoveViewPosition(disappearingItemView,preLayoutInfo);
            return animateRemove(viewHolder);
        }

    }

    protected void adjustRemoveViewPosition(View view,ItemHolderInfo preLayoutInfo){
        int width =View.MeasureSpec.makeMeasureSpec(preLayoutInfo.right-preLayoutInfo.left,View.MeasureSpec.EXACTLY);
        int height =View.MeasureSpec.makeMeasureSpec(preLayoutInfo.bottom-preLayoutInfo.top,View.MeasureSpec.EXACTLY);
        view.measure(width,height);
        view.layout(preLayoutInfo.left, preLayoutInfo.top,
                preLayoutInfo.right,
                preLayoutInfo.bottom);
    }
}
```

Now when updating the complete list using UpdateDataSet(list, true), the original item is no longer misplaced if there is an animation.

Thanks very much to FlexibleAdapter, it's easy to use.

**Comment by davideas** [comment]
@qihongbo, why, was it misplaced? when?
However `viewHolder.isRemoved()` is package private. Cannot compile.

I tried to take preLayoutInfo, but visually nothing change.

**Comment by qihongbo** [comment]
First of all I'm sorry I didn't explain my usage scenarios and requirements.
Second, once again to praise the FlexibleAdapter is a great framework.

Currently the implementation mechanism of FlexibleItemDecoration is for viewType. My requirement is that the same viewType needs different offsets in the same screen. Therefore, FlexibleItemDecoration no longer applies.

If you have Offset configuration function for Item is more flexible. 
So, I defined the ILayoutOffset interface and the ICustomOffset interface:
Public interface ILayoutOffset {
    Rect getLayoutOffset();
    Void setLayoutOffset(int left, int top, int right, int bottom);
    Void setLayoutOffset(Rect rect);
}

Public interface ICustomOffset {
    Rect getCustomOffset();
    Void setCustomOffset(int left, int top, int right, int bottom);
    Void setCustomOffset(Rect rect);
}

The ILayoutOffset interface is for the same viewType and the ICustomOffset interface is for a personalized Offset.

If Item implements ILayoutOffset interface and ICustomOffset interface, then ICustomOffset has a higher priority. I implemented OffsetItemDecoration as follows:
Public class OffsetItemDecoration extends RecyclerView.ItemDecoration {

    @Override
    Public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        Int position = parent.getChildAdapterPosition(view);
        If(position == NO_POSITION){
            Return;
        }

        RecyclerView.Adapter adapter = parent.getAdapter();

        If(adapter instanceof FlexibleAdapter){
            FlexibleAdapter adapterEx = (FlexibleAdapter) adapter;
            IFlexible iFlexible = adapterEx.getItem(position);

            If(iFlexible instanceof ICustomOffset){
                ICustomOffset customOffset = (ICustomOffset)iFlexible;
                Rect rect = customOffset.getCustomOffset();
                If(null != rect){
                    outRect.set(rect);
                    Return;
                }
            }
            If(iFlexible instanceof ILayoutOffset){
                ILayoutOffset viewHolderOffset = (ILayoutOffset)iFlexible;
                Rect rect = viewHolderOffset.getLayoutOffset();
                If(null != rect){
                    outRect.set(rect);
                }
            }
        }
    }

}

Because I have no other special requirements, OffsetItemDecoration is simpler to implement.

At this point, I can already implement different Offsets for different Items of the same viewType in the same screen.

If you call updateDataSet(list, true) to display other items at this time, and use the animation (FlexibleItemAnimator), you will find that the Offset of the deleted Item is invalidated, and there is no Offset between the deleted Items, and the layout becomes quite similar. It does not look good.

So, I had to create FlexibleItemAnimatorEx and put it in android.support.v7.widget package. FlexibleItemAnimatorEx is responsible for restoring the layout between deleted Views.

The above is my little demand, in fact, I want Item to be more flexible...

Thanks again for your framework and hopefully my little implementation will make the FlexibleAdapter even better.

**Comment by davideas** [comment]
@qihongbo Thank you for the compliments.

Can you try to call `invalidateItemDecoration(delay)` at the right point you know?

**Comment by qihongbo** [comment]
@davideas ，
thank you for your reply.
mAdapter.invalidateItemDecorations(delay) does not apply to my needs。

It is recommended that the clear() method of FlexibleItemAnimator be changed from private to protected because of personalization requirements. The following adjustments are recommended in the clear method:
1:delete v.animate().setInterpolator(null).setStartDelay(0); 
2: Add ViewCompat.animate(v).setListener(null);

Thanks

**Comment by davideas** [comment]
@qihongbo, why 1 and 2?

I tried a delete of one and multiple elements and the effect is not good at all.
For instance when we the delete the first or the first 2 subitem all the next elements in the RV (also the others sub elements) moves up one after the other, with the current version all next elements in the RV moves together up in unison.
Same ugly effect when we insert (I tried an Undo).

I won't do let to do this modification of course.
